### PR TITLE
Feature/69/multi contest

### LINKED ISSUE #60: Simplify parameters in ContestRunner constructor
We are passing a lot of arguments to it:

```python
def __init__(self, organizer, teams_root, staff_teams_vs_others_only, include_staff_team, staff_teams_dir, compress_logs,
                 max_steps, no_fixed_layouts,
                 fixed_layouts_file, no_random_layouts, team_names_file,
                 allow_non_registered_students, ignore_file_name_format, www_dir,
                 fixed_layout_seeds=[], random_seeds=[],
                 stats_archive_dir=None, logs_archive_dir=None, replays_archive_dir=None,
                 upload_replays=False, upload_logs=False):
```

This is very error prone, too many options we need to remember their position, etc...

Refactor it to receive a dictionary `options` from where we can then init all these vars, and that's it.

### LINKED ISSUE #69: Split tournament into groups
When we have many teams, one approach could be as follows:

1. First we run a set of sample games to classify each participant into League A or B. A for example are those who won staff medium, and B those who lost.
2. Then we ran 2 tournaments: A play together, and B play together. 

Or there may be other .
=================

However after a meeting we decided to go with the simplest approach. We add an option `--split n` that splits the set of submission teams into `n` groups and runs `n` regular contents.

We need to make sure it stays compatible with the resume feature of contests:

1. Have `tmp-n` for each contest
2. Leave in each `tmp-n` a JSON file with the exact configuraitonof that contest. This should include:
   * Teams that were selected for that game
   * Layouts choosen, fixed and random

Actually 1 and 2 should be done even if no split is done and hence we only have `tmp-1`

The branch to develop this is:

https://github.com/AI4EDUC/pacman-contest-cluster/tree/feature/69/multi-contest

### LINKED ISSUE #72: Resuming failed run doesn't detect all old logs
When resuming a failed run, it chooses pairs of teams with the itertools.combinations function. This may be nondeterministic under some circumstances, so in the previous run it may have had "team-a vs team-b", in the resumed run it plays "team-b vs team-a". This can result in not detecting the old  log files and replaying redundant games

https://github.com/AI4EDUC/pacman-contest-cluster/blob/7caa90e5db0168c2c5e77c4a0045427197c48755/pacman_contest_cluster.py#L935-L948

### PR DESCRIPTION
**Pull request by AndrewPaulChester**
This branch addresses #69 primarily, allowing a MultiContest to contain multiple smaller Contests, which run on a partition of the total team list. 

In doing this, the temporary folder structure has changed and MultiContest settings are now automatically saved to a config file in the tmp folder root to allow easy resuming.

In addition, issues  #60 , and #72 have been resolved, and #65 partially addressed by this branch.

Fixes #69 
Fixes #60 
Fixes #72

### CODE REVIEW

**Comment by ssardina** [comment]
I will start testing and reviewing this. Thanks Andrew.

File: pacman_contest_cluster.py [hunk 1]
```diff
@@ -23,307 +23,754 @@
...
+        self.tmp_contest_dir = os.path.join(TMP_DIR, TMP_CONTEST_DIR)
+
+        # Setup Pacman CTF environment by extracting it from a clean zip file
+        self._prepare_platform(
+            os.path.join(DIR_SCRIPT, CONTEST_ZIP_FILE),
+            settings["fixed_layouts_file"],
+            self.tmp_contest_dir,
+            settings["no_fixed_layouts"],
+            settings["no_random_layouts"],
+            settings.get("fixed_layout_seeds", []),
```
Comment (ssardina) [inline pacman_contest_cluster.py:387]: I changed this as these two seed options may not exist in a run that is not resuming

**Review by ssardina — COMMENTED** [review]
Set the seeds options to default to [] as they may not exist at all in the `settings` dictionary

File: pacman_contest_cluster.py (hunk 1 above)
Comment (ssardina) [reply pacman_contest_cluster.py:387]: This seems to be working but leave it here in case I did not get it and made a mistake for Andrew to double check.

**Comment by ssardina** [comment]
Every contest transfer the whole package from zero. is this because they are different zip files or they are the whole package? If the latter, shouldn't we just take this out and transfer once?

```
Wed, 07 Oct 2020 20:28:59 INFO     FIRST COPYING REQUIRED FILES TO HOSTS....
Wed, 07 Oct 2020 20:28:59 INFO     [START] CORE PACKAGE TRANSFERED TO HOST 118.138.247.84

Wed, 07 Oct 2020 20:28:59 INFO     [START] CORE PACKAGE TRANSFERED TO HOST 118.138.243.84

Wed, 07 Oct 2020 20:28:59 INFO     [START] CORE PACKAGE TRANSFERED TO HOST 118.138.237.185

Wed, 07 Oct 2020 20:28:59 INFO     [START] CORE PACKAGE TRANSFERED TO HOST 118.138.246.177

Wed, 07 Oct 2020 20:28:59 INFO     [START] CORE PACKAGE TRANSFERED TO HOST 118.138.237.124

Wed, 07 Oct 2020 20:28:59 INFO     [START] CORE PACKAGE TRANSFERED TO HOST 118.138.236.74
```

**Comment by ssardina** [comment]
I am doing now a real test on the server with all teams... it seems to be working so far.

File: pacman_contest_cluster.py (hunk 1 above)
Comment (AndrewPaulChester) [reply pacman_contest_cluster.py:387]: Yes, this should do the trick for now. In the medium term (i.e. after the final contest), I would really like to clean up settings a lot, using dataclasses for type safety but also to perform all the validation like this in one place. So you would know that when you have the settings object, all fields are initialised properly (including to defaults), no fields are inconsistent (i.e. giving fixed seeds than there are layouts) etc etc. I've made #79 to track this

**Comment by AndrewPaulChester** [comment]
> Every contest transfer the whole package from zero. is this because they are different zip files or they are the whole package? If the latter, shouldn't we just take this out and transfer once?

They are all the same package, you are right that this is an area for cleaning up. I left it as is to be as backwards compatible as possible, but I should just rewrite the code to just use the MultiContest package for everything - the individual contests shouldn't even need a zip file in their subfolders at all.

I'll have a go at this now, but if you want to merge this PR as is I'll submit a new one for this clean-up.


> I am doing now a real test on the server with all teams... it seems to be working so far.

Results seem good this morning!

**Comment by AndrewPaulChester** [comment]
Fixed this by removing duplication. Not as hard as I had thought:

> Every contest transfer the whole package from zero. is this because they are different zip files or they are the whole package? If the latter, shouldn't we just take this out and transfer once?

**Comment by ssardina** [comment]
Brilliant, so cool.

Yes it ran very well and worked almost out-of-the box. great work! It is very useful to have it now.

File: pacman_contest_cluster.py (hunk 1 above)
Comment (ssardina) [reply pacman_contest_cluster.py]: Just perfect; I see the new issue you created, good!

**Review by ssardina — APPROVED** [review]

**Pull request merged**